File: core/domain/model/shared/HLC.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

package shared

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// HLC is a hybrid logical clock: physical time, a counter, and the device that stamped it
// (offline-sync.md §4.1).
//
// Device clocks are wrong, sometimes by hours. "The latest timestamp wins" therefore lets a device
// with a fast clock outvote every other one permanently, which is why every field change carries
// one of these instead of a bare timestamp: the counter orders changes made within the same
// millisecond, and the device identifier breaks the remaining ties the same way on every device,
// so two clients merging the same pair of changes reach the same result.
//
// The textual form is `<physical>:<counter>:<device>`, with the first two zero-padded so that
// comparing two of them as strings compares them as clocks. The database stores it as text
// (db/schema.sql, change_log.hlc), and an index on that column then sorts in clock order.
type HLC struct {
	// Physical is truncated to milliseconds. Finer resolution would be a promise the counter is
	// there to keep instead, and it would differ between platforms.
	Physical time.Time
	Counter  uint32
	// Device is the origin: a client device, or the installation itself for a server-side change.
	Device string
}

const (
	// hlcPhysicalDigits holds milliseconds since the epoch until the year 2286.
	hlcPhysicalDigits = 13
	// hlcCounterDigits bounds how many changes one device can stamp inside one millisecond before
	// the clock moves on by itself.
	hlcCounterDigits = 5
	hlcMaxCounter    = 99999
)

// NewHLC builds a clock reading. It is the only way to get one with a device that cannot be
// parsed back.
func NewHLC(physical time.Time, counter uint32, device string) (HLC, error) {
	switch {
	case physical.IsZero():
		return HLC{}, ErrInternal.WithDetail("sync.hlc_incomplete")
	case device == "", strings.Contains(device, ":"):
		// A colon in the device would make the text form ambiguous, and an empty device would make
		// two clocks compare equal that are not.
		return HLC{}, ErrInternal.
			WithDetail("sync.hlc_device_malformed").
			WithParams(map[string]string{"device": device})
	case counter > hlcMaxCounter:
		return HLC{}, ErrInternal.WithDetail("sync.hlc_counter_exhausted")
	}
	return HLC{Physical: physical.UTC().Truncate(time.Millisecond), Counter: counter, Device: device}, nil
}

// Tick returns the next reading of this clock for the given wall time.
//
// The physical part only ever moves forward: a wall clock that jumped backwards - an NTP
// correction, a suspended laptop - would otherwise produce a change that sorts before one already
// recorded, and the change would be silently discarded on merge. When time has not moved on, the
// counter does; when the counter is exhausted, the clock borrows a millisecond from the future,
// which is the standard way out and is bounded by how fast a device can produce changes.
func (h HLC) Tick(now time.Time, device string) (HLC, error) {
	now = now.UTC().Truncate(time.Millisecond)

	if h.Physical.IsZero() || now.After(h.Physical) {
		return NewHLC(now, 0, device)
	}
	if h.Counter >= hlcMaxCounter {
		return NewHLC(h.Physical.Add(time.Millisecond), 0, device)
	}
	return NewHLC(h.Physical, h.Counter+1, device)
}

// String is the stored and transmitted form.
func (h HLC) String() string {
	if h.Physical.IsZero() {
		return ""
	}
	return fmt.Sprintf("%0*d:%0*d:%s",
		hlcPhysicalDigits, h.Physical.UnixMilli(), hlcCounterDigits, h.Counter, h.Device)
}

// IsZero reports the absent reading.
func (h HLC) IsZero() bool { return h.Physical.IsZero() }

// ParseHLC reads the textual form. It is what a client's mutation arrives as, so an unusable value
// is input rather than a defect - the caller decides what to do with a client that sends nonsense
// (offline-sync.md §3.2).
func ParseHLC(raw string) (HLC, error) {
	malformed := ErrValidation.
		WithDetail("sync.hlc_malformed").
		WithParams(map[string]string{"value": raw})

	physical, rest, found := strings.Cut(raw, ":")
	if !found {
		return HLC{}, malformed
	}
	counter, device, found := strings.Cut(rest, ":")
	if !found || device == "" {
		return HLC{}, malformed
	}

	milliseconds, err := strconv.ParseInt(physical, 10, 64)
	if err != nil || milliseconds < 0 {
		return HLC{}, malformed
	}
	ticks, err := strconv.ParseUint(counter, 10, 32)
	if err != nil || ticks > hlcMaxCounter {
		return HLC{}, malformed
	}

	return HLC{
		Physical: time.UnixMilli(milliseconds).UTC(),
		Counter:  uint32(ticks),
		Device:   device,
	}, nil
}

// Compare orders two readings: physical time first, then the counter, then the device.
//
// The device is part of the order rather than a tie-break left to chance. Two devices that stamped
// a change in the same millisecond with the same counter have to be ordered the same way
// everywhere, or the two of them converge on different results - which is the one failure a
// merge rule must not have.
func (h HLC) Compare(other HLC) int {
	if !h.Physical.Equal(other.Physical) {
		if h.Physical.Before(other.Physical) {
			return -1
		}
		return 1
	}
	if h.Counter != other.Counter {
		if h.Counter < other.Counter {
			return -1
		}
		return 1
	}
	return strings.Compare(h.Device, other.Device)
}

// After reports whether this reading wins a last-writer-wins comparison against the other one.
func (h HLC) After(other HLC) bool { return h.Compare(other) > 0 }

File: packages/sync-engine/src/hlc.ts
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

// The hybrid logical clock, on this side of the wire (offline-sync.md §4.1).
//
// Device clocks are wrong, sometimes by hours. Every field change a device pushes therefore
// carries one of these rather than a bare timestamp: physical time, a counter that orders changes
// made within one millisecond, and the device that stamped it, which breaks the remaining ties the
// same way everywhere. The textual form is the server's - `core/domain/model/shared/HLC.go` -
// `<physical>:<counter>:<device>`, the first two zero-padded so that comparing two readings as
// strings compares them as clocks.
//
// What this module decides is `Tick`'s rule, and only that: the physical part never moves
// backwards. A wall clock that jumped back - a suspended laptop, a correction - would otherwise
// stamp a change that sorts before one already made, and the server would discard it on merge.
// When time has not moved on, the counter does; when the counter is exhausted, the clock borrows a
// millisecond from the future. The physical time comes from the `Clock` port and never from the
// machine directly, which is what lets a test fix it.

import type { Clock } from './ports.ts';

/** Milliseconds since the epoch, padded: enough digits until the year 2286. */
const PHYSICAL_DIGITS = 13;
/** How many changes one device can stamp inside one millisecond before the clock moves by itself. */
const COUNTER_DIGITS = 5;
export const HLC_MAX_COUNTER = 99_999;

export interface HlcReading {
  /** Milliseconds since the epoch, truncated to the millisecond. */
  readonly physical: number;
  readonly counter: number;
  readonly device: string;
}

/** The stored and transmitted form, as the server parses it. */
export function formatHlc({ physical, counter, device }: HlcReading): string {
  return `${String(physical).padStart(PHYSICAL_DIGITS, '0')}:${String(counter).padStart(COUNTER_DIGITS, '0')}:${device}`;
}

/**
 * Reads the textual form back, or nothing for a value the server would refuse too
 * (`sync.hlc_malformed`). A reading this client did not mint is still worth parsing: the server's
 * records carry one, and a device compares them to its own.
 */
export function parseHlc(raw: string): HlcReading | undefined {
  const first = raw.indexOf(':');
  if (first < 0) return undefined;
  const second = raw.indexOf(':', first + 1);
  if (second < 0) return undefined;
  const physical = Number(raw.slice(0, first));
  const counter = Number(raw.slice(first + 1, second));
  const device = raw.slice(second + 1);
  if (!Number.isInteger(physical) || physical < 0) return undefined;
  if (!Number.isInteger(counter) || counter < 0 || counter > HLC_MAX_COUNTER) return undefined;
  if (device === '') return undefined;
  return { physical, counter, device };
}

/** Physical time first, then the counter, then the device - the server's `Compare`. */
export function compareHlc(a: HlcReading, b: HlcReading): number {
  if (a.physical !== b.physical) return a.physical < b.physical ? -1 : 1;
  if (a.counter !== b.counter) return a.counter < b.counter ? -1 : 1;
  return a.device < b.device ? -1 : a.device > b.device ? 1 : 0;
}

/**
 * The next reading after `last`, at the clock's wall time: the server's `Tick`, with the same
 * three branches in the same order.
 */
export function tickHlc(last: HlcReading | undefined, now: number, device: string): HlcReading {
  const wall = Math.max(0, Math.trunc(now));
  if (last === undefined || wall > last.physical) return { physical: wall, counter: 0, device };
  if (last.counter >= HLC_MAX_COUNTER) return { physical: last.physical + 1, counter: 0, device };
  return { physical: last.physical, counter: last.counter + 1, device };
}

/**
 * One device's clock: the last reading it made, and the next one on demand. The queue stamps
 * every mutation from here; the reading is also what a device compares a server's record
 * against, which is why `observe` exists - a device that has seen a later reading must not stamp
 * an earlier one.
 */
export class HybridClock {
  readonly #clock: Clock;
  readonly #device: string;
  #last: HlcReading | undefined;

  constructor(clock: Clock, device: string, last?: HlcReading) {
    if (device === '' || device.includes(':')) {
      throw new TypeError('a device identifier has to be non-empty and free of colons - it is a segment of the clock');
    }
    this.#clock = clock;
    this.#device = device;
    this.#last = last;
  }

  /** The reading last handed out, for a caller that keeps it in the store across a reload. */
  get last(): HlcReading | undefined {
    return this.#last;
  }

  /** The next reading, and it is remembered. */
  next(): HlcReading {
    this.#last = tickHlc(this.#last, this.#clock.now(), this.#device);
    return this.#last;
  }

  /** The next reading, formatted - what a mutation carries. */
  stamp(): string {
    return formatHlc(this.next());
  }

  /**
   * A reading seen from elsewhere - a record the server sent. The clock moves past it, so that
   * the next stamp sorts after everything this device has seen, whatever its wall clock says.
   */
  observe(reading: HlcReading): void {
    if (this.#last === undefined || compareHlc(reading, this.#last) > 0) {
      this.#last = { physical: reading.physical, counter: reading.counter, device: this.#device };
    }
  }
}
